packages/framework-core/src/services/event-store.ts

```diff
@@ -17,90 +17,72 @@ const originOfTime = new Date(0).toISOString() // Unix epoch
 
     if (pendingEvents.length <= 0) {
       return latestSnapshotEnvelope
     } else {
-      logger.debug(
-        `[EventStore#fetchEntitySnapshot] Looking for the reducer for entity ${entityName} with ID ${entityID}`
-      )
+      logger.debug(`Looking for the reducer for entity ${entityName} with ID ${entityID}`)
       let newEntitySnapshot = latestSnapshotEnvelope
       for (const pendingEvent of pendingEvents) {
-        newEntitySnapshot = await this.entityReducer(newEntitySnapshot, pendingEvent)
+        // We double check that what we are reducing is an event
+        if (pendingEvent.kind === 'event') {
+          newEntitySnapshot = await this.entityReducer(newEntitySnapshot, pendingEvent)
+        }
       }
 
-      logger.debug(
-        `[EventStore#fetchEntitySnapshot] Reduced new snapshot for entity ${entityName} with ID ${entityID}: `,
-        newEntitySnapshot
-      )
-
-      return newEntitySnapshot
-    }
-  }
-
-  public async calculateAndStoreEntitySnapshot(
-    entityName: string,
-    entityID: UUID,
-    pendingEnvelopes: Array<EventEnvelope>
-  ): Promise<Array<EventEnvelope>> {
-    const logger = getLogger(this.config, 'EventStore#calculateAndStoreEntitySnapshot')
-    logger.debug('Processing events: ', pendingEnvelopes)
-    logger.debug(`Fetching snapshot for entity ${entityName} with ID ${entityID}`)
-    const latestSnapshotEnvelope = await this.loadLatestSnapshot(entityName, entityID)
-    const snapshotsPerNewEntityId: Record<string, EventEnvelope> = {}
-    if (latestSnapshotEnvelope) {
-      snapshotsPerNewEntityId[latestSnapshotEnvelope.entityID.toString()] = latestSnapshotEnvelope
-    }
+      if (newEntitySnapshot?.entityID !== entityID) {
+        logger.debug(
+          `Migrated entity ${entityName} with previous ID ${entityID} to ${newEntitySnapshot?.typeName} with the new ID ${newEntitySnapshot?.entityID}`,
+          newEntitySnapshot
+        )
+      } else {
+        logger.debug(`Reduced new snapshot for entity ${entityName} with ID ${entityID}: `, newEntitySnapshot)
+      }
 
-    logger.debug(
-      `[EventStore#calculateAndStoreEntitySnapshot] Looking for the reducer for entity ${entityName} with ID ${entityID}`
-    )
-    for (const pendingEvent of pendingEnvelopes) {
-      const newEntitySnapshot = await this.entityReducer(
-        snapshotsPerNewEntityId[pendingEvent.entityID.toString()],
-        pendingEvent
-      )
-      snapshotsPerNewEntityId[newEntitySnapshot.entityID.toString()] = newEntitySnapshot
-    }
+      if (!newEntitySnapshot) {
+        logger.debug('No snapshot was fetched, returning')
 
-    logger.debug(
-      `[EventStore#calculateAndStoreEntitySnapshot] Reduced new snapshots for entity ${entityName} with ID ${entityID}: `,
-      snapshotsPerNewEntityId
-    )
+        return newEntitySnapshot
+      }
 
-    const snapshotsForNewEntityIds = Object.values(snapshotsPerNewEntityId)
+      await this.storeSnapshot(newEntitySnapshot)
```

Comment:
Correct me if I'm wrong, but if we persist here then we can have concurrency issues. See the following diagram:

![booster_concurrency_3 001](https://user-images.githubusercontent.com/1274959/216042789-6c748530-fd6d-4915-83d5-fd2b9cd18ecd.jpeg)

Do you thinks this use case is possible? If so, will Booster lost the event number 3 in this case?

Reply:
This is fun, I've always assumed that snapshots were stored using the last event `createdAt` value as their sort key, but reviewing the code it seems that we're storing them as they come with the current timestamp. I think that the current behavior is a mistake. It's okay to store the time of actual creation for reference, but in my opinion, the sortKey should be exactly the same as the latest processed event's.

So yeah, with the current behavior what you're describing is fully right, but I think we should change the way snapshots are stored to use the last event's sortKey as their sortKey. This way, the `loadLatestSnapshot` method, which sorts the snapshots by this field and takes the latest, will always get the snapshot that includes the latest event, no matter in which order it was inserted in the database.

Another thing that could happen with the current implementation is that when you're storing a snapshot that already exists in the database (because two threads calculated the same one after reading the same list of events), the keys don't match because each of them will have their own timestamp, they'll end up duplicated but with different identifiers. If you try to persist an already existing snapshot, it should be ignored.

Reply:
Makes sense. I just realized that `createdAt` is not only the current timestamp when the entity is built on `core` but it is overridden for each Provider before persisting the entity with a new `current timestamp`.

About your last comment:
```
Another thing that could happen with the current implementation is that when you're storing a snapshot that already exists in the database (because two threads calculated the same one after reading the same list of events), the keys don't match because each of them will have their own timestamp, they'll end up duplicated but with different identifiers. If you try to persist an already existing snapshot, it should be ignored.
```

For Azure, there is only a partition key defined. AFAIK there is not any unique constraint in the container. For Azure, if you want uniqueness, you need to define it explicitly.

Reply:
Gotcha, I'm working on fixing this here: https://github.com/boostercloud/booster/pull/1331